src/lib/utils/bitvector.h

```diff
@@ -0,0 +1,1296 @@
+                  m_block(blocks[block_index(pos)]), m_mask(one << block_offset(pos)) {}
+
+         public:
+            bitref_base() = delete;
+            bitref_base(const bitref_base&) noexcept = default;
+            bitref_base(bitref_base&&) noexcept = default;
+            bitref_base& operator=(const bitref_base&) = delete;
+            bitref_base& operator=(bitref_base&&) = delete;
+
+            ~bitref_base() = default;
+
+         public:
+            constexpr operator bool() const noexcept { return is_set(); }
+
+            constexpr bool is_set() const noexcept { return (m_block & m_mask) > 0; }
+
+            template <std::integral T>
+            constexpr T as() const noexcept {
+               return static_cast<T>(is_set());
+            }
+
+         protected:
+            BlockT& m_block;  // NOLINT(*-non-private-member-variables-in-classes)
+            BlockT m_mask;    // NOLINT(*-non-private-member-variables-in-classes)
+      };
+
+   public:
+      /**
+       * Wraps a constant reference into the bitvector. Bit can be accessed
+       * but not modified.
+       */
+      template <typename BlockT>
+      class bitref final : public bitref_base<BlockT> {
+         public:
+            using bitref_base<BlockT>::bitref_base;
+      };
+
+      /**
+       * Wraps a modifiable reference into the bitvector. Bit may be accessed
+       * and modified (e.g. flipped or XOR'ed).
+       */
+      template <typename BlockT>
+         requires(!std::is_const_v<BlockT>)
+      class bitref<BlockT> : public bitref_base<BlockT> {
+         public:
+            using bitref_base<BlockT>::bitref_base;
+
+            ~bitref() = default;
+            bitref(const bitref&) noexcept = default;
+            bitref(bitref&&) noexcept = default;
+
+            constexpr bitref& set() noexcept {
+               this->m_block |= this->m_mask;
+               return *this;
+            }
+
+            constexpr bitref& unset() noexcept {
+               this->m_block &= ~this->m_mask;
+               return *this;
+            }
+
+            constexpr bitref& flip() noexcept {
+               this->m_block ^= this->m_mask;
+               return *this;
+            }
+
+            // NOLINTBEGIN
+
+            constexpr bitref& operator=(bool bit) noexcept { return assign(bit); }
+
+            constexpr bitref& operator=(const bitref& bit) noexcept { return assign(bit); }
+
+            constexpr bitref& operator=(bitref&& bit) noexcept { return assign(bit); }
+
+            // NOLINTEND
+
+            constexpr bitref& operator&=(bool other) noexcept { return assign(this->is_set() && other); }
+
+            constexpr bitref& operator|=(bool other) noexcept { return assign(this->is_set() || other); }
+
+            constexpr bitref& operator^=(bool other) noexcept { return assign(this->is_set() ^ other); }
+
+         private:
+            constexpr bitref& assign(bool bit) noexcept { return (bit) ? set() : unset(); }
```

Comment:
As discussed, [DATA](https://github.com/Fraunhofer-AISEC/DATA) identified this line as a leakage during our SCA review.

Due to the `?` operator, a control-flow difference is observed based on the boolean input `bit` variable.

The `assign()` routine is used within the [`push_back()`](https://github.com/randombit/botan/pull/3883/files#diff-011c69221587ab2822aae9e7e1994a2cf3d35364206c4e65e5882362ffee7e8eR531-R535) routine, which in turn is used within the [`decode()`](https://github.com/randombit/botan/pull/3883/files#diff-9b49912d2f52d848b3d823202bd961beea745c8516271cba858747d1ed029210R109) routine. This may allow an adversary to observe the error vector `e` and, hence recover the shared secret.

We suggest to perform both the `set()` and `unset()` functions with the input as a mask, for example:
```
         private:
            constexpr bitref& assign(bool bit) noexcept {
                const block_type assign_mask = 0 - static_cast<block_type>(bit);
                this->m_block |=  (this->m_mask &  assign_mask);
                this->m_block &= ~(this->m_mask & ~assign_mask);
                return *this;
            }
```

In our case, this results in the following instructions - without any conditional branch based on the input:
```
[ ... ] 
                                         const block_type assign_mask = 0 - static_cast<block_type>(bit);
41d397: 41 f7 dc                   neg %r12d
[ ... ]
                                         this->m_block |= (this->m_mask & assign_mask);
41d3ab: 44 89 e1                   mov %r12d,%ecx
41d3ae: 21 c1                      and %eax,%ecx
                                         this->m_block &= ~(this->m_mask & ~assign_mask);
41d3b0: f7 d0                      not %eax
                                         this->m_block |= (this->m_mask & assign_mask);
41d3b2: 0a 0a                      or (%rdx),%cl
                                         this->m_block &= ~(this->m_mask & ~assign_mask);
41d3b4: 44 09 e0                   or %r12d,%eax
41d3b7: 21 c8                      and %ecx,%eax
[ ... ]
```

Reply:
Thanks for your analysis and your report! I applied your fix using Botan's constant-time helper class (7a0fe59)